src/server.py

```diff
@@ -33,94 +68,155 @@ def server():
     try:
         while True:
             conn, addr = serv_sock.accept()
-            request = []
-            while True:
-                part = conn.recv(BUFFER_LENGTH)
-                request.append(part)
-                # conn.sendall(part)
-                if len(part) < BUFFER_LENGTH:
-                    break
-            request = b''.join(request)
-            print('Request received:')
-            print(request.decode('utf-8'))
-            try:
-                uri = parse_request(request)
-                response = response_ok()
-            except ValueError as e:
-                response = response_error(*e.args)
-            conn.sendall(response)
-            conn.close()
-            time.sleep(0.01)
-    except KeyboardInterrupt:
-        pass
+            http_server(conn, addr)
+    except Exception as e:
+        print(e.msg)
     finally:
-        try:
-            conn.close()
-        except NameError:
-            pass
         print('\nShutting down the server...\n')
         serv_sock.close()
         sys.exit()
 
 
-def parse_request(request):
-    """TOOD: Docstring."""
-    # only accept GET
-    # only HTTP/1.1
-    # validate that proper Host: header was specified
-    # other requests raise appropriate Python error
-    # if no conditions arise, should return the URI
-    method = b''
-    uri = b''
-    protocol = b''
-    headers = b''
+def http_server(conn, addr):
+    """Take HTTP requests and return appropriate HTTP response."""
+    with closing(conn):
+        request_parts = []
+        while True:
+            part = conn.recv(BUFFER_LENGTH)
+            request_parts.append(part)
+            if len(part) < BUFFER_LENGTH:
+                break
+        # Immediately decode all of incoming request into unicode.
+        # In future, may need to check Content-type of incoming request?
```

Comment:
absolutely yes.  If the incoming request has a body, then you'd want to check both the content type and the encoding before decoding.